# Exploring eBay Car Sales Data

This is a guided projected from dataquest.io completed as part of the Data Science with Python course.

Thius project will clean and explore 50,000 rows of data originally scraped from eBay. I will primarily be using the Pandas library to analyse the dataset.

In [3]:
# firstly import the required libraries
import numpy as np
import pandas as pd

In [4]:
# read the data
autos = pd.read_csv('data/autos.csv',encoding='Latin-1')

In [5]:
# print summaries
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [6]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset includes 20 columns and 50,00 rows.

- The columns: `vehicleType, gearbox, model, fuelType and notRepairedDamaged` all contain null values
- `dateCrawled, dateCreated and lastSeen` are all dates stored as strings
- `price and odometer` are also numeric columns stored as strings
- the column names are camelcase rather than the prefered snakecase

In [7]:
# Firstly amend the column names, start by getting the names
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [8]:
# create list with prefered column names
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_pictures', 'postal_code',
       'last_seen']
# apply new columns to df
autos.columns = new_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Changes were made to make the columns as descriptive as possible, whilst adhering to the snakecase convention.

In [9]:
# get descriptive stats for all columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

- Columns with mostly one value that could be dropped: `seller, offer_type, fuel_type, unrepaired_damage`
- Columns with numbers stored as text: `price, odometer`

In [10]:
# convert price and odometer columns
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(int)
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].astype(int)
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)
autos.describe()

D:\TEMP\ipykernel_2488\1577337605.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$","")


price  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nr_pictures   postal_code  
count        50000.000000      50000.0  50000.000000  
mean             5.723360          0.0  50813.627300  
std              3.711984          0.0  25779.747957  
min              0.000000          0.0   1067.000000  
25%              3.000000          0.0  30451.000000  
50%              6.000000          0.0  49577.000000  
75%              9.000000          0.0  71540.000000  
max             12.000000          0.0  99998.000000

In [11]:
# explore price and odometer_km further
print("Price")
print(autos["price"].unique().shape)
print(autos["price"].value_counts().sort_index(ascending=False).head())
print(autos["price"].value_counts().sort_index().head())
print("\n")
print("Odometer km")
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].value_counts().sort_index(ascending=False).head())
print(autos["odometer_km"].value_counts().sort_index().head())

Price
(2357,)
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price, dtype: int64
0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64


Odometer km
(13,)
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64
5000     967
10000    264
20000    784
30000    789
40000    819
Name: odometer_km, dtype: int64


In [12]:
# remove price outliers
autos = autos[autos["price"].between(500,100000)]
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  45058.000000       45058.000000  45058.000000   45058.000000   
mean    6177.407342        2005.063918    121.336788  125364.752097   
std     7636.932471          89.689852    205.622294   39542.959058   
min      500.000000        1000.000000      0.000000    5000.000000   
25%     1500.000000        2000.000000     75.000000  100000.000000   
50%     3500.000000        2004.000000    110.000000  150000.000000   
75%     7900.000000        2008.000000    150.000000  150000.000000   
max    99900.000000        9999.000000  17700.000000  150000.000000   

       registration_month  nr_pictures   postal_code  
count        45058.000000      45058.0  45058.000000  
mean             5.907675          0.0  51283.869169  
std              3.633063          0.0  25700.705805  
min              0.000000          0.0   1067.000000  
25%              3.000000          0.0  30952.000000  
50%              6.000000          0.0  50226.000000  
75%              9.000000          0.0  72156.750000  
max             12.000000          0.0  99998.000000

From the `price` column outliers were identified. Prices were seen in the millions and also next to zero, this seems unrealistic for a car sales website. Prices were filtered to between 500 and 10,000.

Data in the `odometer_km` column is skewed heavily, with almost half the rows having 150,000 km. In can be assumed this is the case because it is the maximum entry on the website. As 150,000 km does not seem unrealistic the rows will not be removed. 

Next the date columns `date_crawled, last_seen and ad_created` will be explored

In [13]:
# display the distribution of the date columns
print("date crawled")
print(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index())
print("ad created")
print(autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index())
print("last_seen")
print(autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

date crawled
2016-03-05    0.025545
2016-03-06    0.014137
2016-03-07    0.036220
2016-03-08    0.033157
2016-03-09    0.032913
2016-03-10    0.032736
2016-03-11    0.033046
2016-03-12    0.037330
2016-03-13    0.015536
2016-03-14    0.036331
2016-03-15    0.034001
2016-03-16    0.029362
2016-03-17    0.031160
2016-03-18    0.012850
2016-03-19    0.034777
2016-03-20    0.038084
2016-03-21    0.037729
2016-03-22    0.032935
2016-03-23    0.032403
2016-03-24    0.028985
2016-03-25    0.031093
2016-03-26    0.032625
2016-03-27    0.031160
2016-03-28    0.034822
2016-03-29    0.033290
2016-03-30    0.033290
2016-03-31    0.031670
2016-04-01    0.033867
2016-04-02    0.035798
2016-04-03    0.038817
2016-04-04    0.036642
2016-04-05    0.013161
2016-04-06    0.003174
2016-04-07    0.001354
Name: date_crawled, dtype: float64
ad created
2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
                ...   
2016-04-03    0.03903

`date_crawled`: roughly even distribution of dates from 5 Mar 2016 to Apr 2016

`ad_created`:  dates ranging from 10 Aug 2015 to 7 Apr 2016, evenly distributed

`last_seen`: same date range as `date_crawled` but more than 40% of occurances in the last 3 days. 

In [14]:
# stats for reg year column
autos["registration_year"].describe()

count    45058.000000
mean      2005.063918
std         89.689852
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Min value in `registration_year` is 1000 and the max is 9999. Both incorrect values for car years

In [15]:
autos = autos.loc[autos["registration_year"].between(1900,2016)]

In [16]:
print(autos["registration_year"].value_counts(normalize=True))

2005    0.066168
2000    0.062702
2004    0.061986
2006    0.061593
2003    0.061385
          ...   
1927    0.000023
1943    0.000023
1931    0.000023
1948    0.000023
1951    0.000023
Name: registration_year, Length: 78, dtype: float64


after restricting `registration_year` to 1900 (when cars invented) to 2016 (when data produced) a much more reasonable subset can be seen.

Next i will explore the `brand` column to see which brand had the highest mean price.

In [17]:
# get list of top 10 most popuular brands by number of listings
brand_values = autos["brand"].value_counts().sort_values(ascending=False).head(10)
brands = brand_values.index
print(brands)

Index(['volkswagen', 'bmw', 'mercedes_benz', 'opel', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [18]:
# create dictionary to agregate brand by mean price
brand_mean_price = {}
for brand in brands:
    mean_price = autos.loc[autos["brand"] == brand, "price"].mean()
    brand_mean_price[brand] = mean_price
bmp_series = pd.Series(brand_mean_price)
print(bmp_series.sort_values(ascending=False))

audi             9571.457398
mercedes_benz    8666.677208
bmw              8447.069880
volkswagen       5783.622985
seat             4810.883871
ford             4247.120482
opel             3394.039568
peugeot          3360.920597
fiat             3256.152110
renault          2819.059411
dtype: float64


Audi, with 9571, has the highest mean price of the top 10 brands, followed by BMW and Mercedes.

Renault, with 2819, has the lowest mean price of the top 10 brands.

Next, I will do the same with `odometer_km`

In [19]:
# create dictionary to agregate brand by mean milage
brand_mean_milage = {}
for brand in brands:
    mean_milage = autos.loc[autos["brand"] == brand, "odometer_km"].mean()
    brand_mean_milage[brand] = mean_milage
bmm_series = pd.Series(brand_mean_milage)
print(bmm_series.sort_values(ascending=False))

bmw              132928.714859
mercedes_benz    131083.126271
audi             128941.326531
volkswagen       128234.749455
opel             128012.422360
renault          126351.209253
peugeot          126073.113208
ford             123520.552799
seat             120058.064516
fiat             114416.094210
dtype: float64


BMW has the highest milage of the top 10 brands, Fiat has the lowest

In [20]:
d = {'mean_price':bmp_series,'mean_odom_km':bmm_series, 'ad_count':brand_values}
top_brands_df = pd.DataFrame(d)
print(top_brands_df)

                mean_price   mean_odom_km  ad_count
volkswagen     5783.622985  128234.749455      9180
bmw            8447.069880  132928.714859      4980
mercedes_benz  8666.677208  131083.126271      4427
opel           3394.039568  128012.422360      4347
audi           9571.457398  128941.326531      3920
ford           4247.120482  123520.552799      2822
renault        2819.059411  126351.209253      1902
peugeot        3360.920597  126073.113208      1272
fiat           3256.152110  114416.094210      1019
seat           4810.883871  120058.064516       775


### Top 10 Brands

- Most popular `volkswagen`
- Least popular ` seat`
- Most expensive `audi`
- Cheapest `renault`
- Highest milage `BMW`
- Lowest milage `fiat`